Google.PowerShell/Common/ActiveUserToken.cs

```diff
@@ -12,37 +14,131 @@ public class ActiveUserToken
         /// <summary>
         /// The current active user that corresponds to this token.
         /// </summary>
-        private string activeUser = CloudSdkSettings.GetSettingsValue("account");
+        private string activeUser;
 
-        /// <summary>Gets or sets the access token issued by the authorization server.</summary>
-        [Newtonsoft.Json.JsonPropertyAttribute("access_token")]
-        public string AccessToken { get; set; }
+        /// <summary>
+        /// This function will return the current active user.
+        /// </summary>
+        internal Func<string> getActiveUser;
 
-        /// <summary>Gets or sets the lifetime in seconds of the access token.</summary>
-        [Newtonsoft.Json.JsonPropertyAttribute("expires_in")]
-        public Nullable<long> ExpiresInSeconds { get; set; }
+        /// <summary>The access token issued by the authorization server.</summary>
+        public string AccessToken { get; internal set; }
 
-        /// <summary>The date and time that this token was issued. <remarks>
+        /// <summary>
+        /// The date and time that this token was issued.
+        /// This is set in UTC time.
         /// It should be set by the CLIENT after the token was received from the server.
-        /// </remarks> 
         /// </summary>
-        public DateTime Issued { get; set; }
+        public DateTime ExpiredTime { get; internal set; }
+
+        /// <summary>
+        /// Constructs a new token by parsing the userTokenJson.
+        /// GetActiveUser will be using CloudSdkSettings.GetSettingsValue("account").
+        /// </summary>
+        public ActiveUserToken(string userTokenJson) : this(userTokenJson, () => CloudSdkSettings.GetSettingsValue("account"))
+        {
+        }
+
+        /// <summary>
+        /// Construct a new token by parsing the userTokenJson.
+        /// The activeUser will be computed by calling getActiveUser.
+        /// </summary>
+        public ActiveUserToken(string userCredentialJson, Func<string> getActiveUser)
+        {
+            this.getActiveUser = getActiveUser;
+            activeUser = this.getActiveUser();
+
+            JToken parsedCredentialJson = JObject.Parse(userCredentialJson);
+            JToken accessTokenJson = parsedCredentialJson.SelectToken("access_token");
+
+            if (accessTokenJson == null || accessTokenJson.Type != JTokenType.String)
+            {
+                throw new InvalidDataException("Credential Json should contain access token key.");
+            }
+
+            AccessToken = accessTokenJson.Value<string>();
+
+            JToken tokenExpiryJson = parsedCredentialJson.SelectToken("token_expiry");
+
+            // Token from GCE does not have expiry.
```

Comment:
"Token from GCE" isn't quite a technical definition. I believe it is "Service account credentials do not expire."